File: related_scripts/verification/step_E/benchmark_varcorr.py
"""Benchmark Python variance correction at production scale."""
import sys
import time
from pathlib import Path

import numpy as np

sys.path.insert(0, str(Path(__file__).resolve().parents[3] / "src"))
from caz.gridding import variance_correction, _movmean_2d, _movstd_2d

def benchmark(n_cells, n_time, window=30, label=""):
    print(f"--- {label} ({n_cells}x{n_time}) ---")
    np.random.seed(42)
    pcr = np.random.randn(n_cells, n_time)
    esm = np.random.randn(n_cells, n_time)

    # movmean
    t0 = time.perf_counter()
    pcr_mm = _movmean_2d(pcr, window)
    t_mm = time.perf_counter() - t0
    print(f"  movmean:  {t_mm:.3f} s")

    # movstd
    t0 = time.perf_counter()
    pcr_ms = _movstd_2d(pcr - pcr_mm, window)
    t_ms = time.perf_counter() - t0
    print(f"  movstd:   {t_ms:.3f} s")

    # full variance correction
    t0 = time.perf_counter()
    adj = variance_correction(pcr, esm, window=window)
    t_full = time.perf_counter() - t0
    print(f"  Full varcorr: {t_full:.3f} s")
    ram_gb = n_cells * n_time * 8 * 8 / 1e9
    print(f"  Peak RAM est: {ram_gb:.1f} GB\n")
    return t_mm, t_ms, t_full

if __name__ == "__main__":
    print("=== Variance Correction Benchmark (Python, cumsum-optimized) ===\n")

    results = {}
    for label, nc, nt in [
        ("Small", 50, 200),
        ("Medium", 18030, 1500),
    ]:
        results[label] = benchmark(nc, nt, label=label)

    # Full grid only if enough memory
    try:
        results["Full"] = benchmark(180901, 1500, label="Full")
    except MemoryError:
        print("--- Full (180901x1500) --- SKIPPED: MemoryError\n")

    print("Done.")

File: related_scripts/verification/step_E/benchmark_varcorr.m
% Benchmark variance correction at production scale
% Compare MATLAB movmean/movstd speed on realistic grid sizes.

fprintf('=== Variance Correction Benchmark (MATLAB) ===\n\n');

rng(42);
n_year_mov = 30;

sizes = {
    'Small (50x200)',      50,   200;
    'Medium (18030x1500)', 18030, 1500;
    'Full (180901x1500)',  180901, 1500;
};

for k = 1:size(sizes, 1)
    label   = sizes{k, 1};
    n_cells = sizes{k, 2};
    n_time  = sizes{k, 3};

    fprintf('--- %s ---\n', label);

    if n_cells > 100000
        fprintf('  (Full grid: %.1f GB RAM needed, skipping if insufficient)\n', ...
            n_cells * n_time * 8 * 8 / 1e9);
        try
            pcr = randn(n_cells, n_time);
            esm = randn(n_cells, n_time);
        catch ME
            fprintf('  SKIPPED: %s\n\n', ME.message);
            continue
        end
    else
        pcr = randn(n_cells, n_time);
        esm = randn(n_cells, n_time);
    end

    % Warm up
    if n_cells <= 18030
        tmp = movmean(pcr(1:min(100,n_cells), :), n_year_mov, 2, 'omitnan'); %#ok<NASGU>
    end

    % Benchmark movmean
    tic;
    pcr_mm = movmean(pcr, n_year_mov, 2, 'omitnan');
    t_mm = toc;
    fprintf('  movmean:  %.3f s\n', t_mm);

    % Benchmark movstd
    tic;
    pcr_ms = movstd(pcr - pcr_mm, n_year_mov, 0, 2, 'omitnan');
    t_ms = toc;
    fprintf('  movstd:   %.3f s\n', t_ms);

    % Benchmark full variance correction
    tic;
    pcr_mm2 = movmean(pcr, n_year_mov, 2, 'omitnan');
    esm_mm2 = movmean(esm, n_year_mov, 2, 'omitnan');
    std_c = movstd(esm - esm_mm2, n_year_mov, 0, 2, 'omitnan') ./ ...
            movstd(pcr - pcr_mm2, n_year_mov, 0, 2, 'omitnan');
    std_c(~isfinite(std_c)) = 1;
    adjusted = pcr_mm2 + (pcr - pcr_mm2) .* std_c;
    t_full = toc;
    fprintf('  Full varcorr: %.3f s\n', t_full);
    fprintf('  Peak RAM est: %.1f GB\n\n', n_cells * n_time * 8 * 8 / 1e9);
end

fprintf('Done.\n');
